cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,163 @@
+use std::iter;
+use tracing::trace;
+
+/// The secret hash sharing protocol works in the following way:
+///
+/// - Dialer (Alice) writes the `Message` to the substream.
+/// - Listener (Bob) reads the `Message` from the substream.
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Clone, Copy, Deserialize, Debug, Serialize, PartialEq)]
+pub struct Message {
+    swap_id: SwapId,
+    secret_hash: SecretHash,
+}
+
+/// Represents a prototype for an upgrade to handle the sender side of the
+/// secret hash sharing protocol.  Config contains the `Message`, once the
+/// outbound upgrade is complete peer node has been sent the message.
+#[derive(Clone, Copy, Debug)]
+pub struct OutboundProtocolConfig {
+    msg: Message,
+}
+
+impl UpgradeInfo for OutboundProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> OutboundUpgrade<C> for OutboundProtocolConfig
+where
+    C: AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = ();
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, mut socket: C, info: Self::Info) -> Self::Future {
+        trace!(
+            "Upgrading outbound connection: {}",
+            String::from_utf8_lossy(info)
+        );
+        Box::pin(async move {
+            let bytes = serde_json::to_vec(&self.msg).unwrap();
+            upgrade::write_one(&mut socket, &bytes).await.unwrap();
+
+            Ok(())
+        })
+    }
+}
+
+/// Represents a prototype for an upgrade to handle the receiver side of the
+/// secret hash sharing protocol.
+#[derive(Clone, Copy, Debug)]
+pub struct InboundProtocolConfig;
+
+impl Default for InboundProtocolConfig {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+impl UpgradeInfo for InboundProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> InboundUpgrade<C> for InboundProtocolConfig
+where
+    C: AsyncRead + Unpin + Send + 'static,
+{
+    type Output = Message;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_inbound(self, mut socket: C, info: Self::Info) -> Self::Future {
+        trace!(
+            "Upgrading inbound connection: {}",
+            String::from_utf8_lossy(info)
+        );
+        Box::pin(async move {
+            let message = upgrade::read_one(&mut socket, 1024).await?;
+            let mut de = serde_json::Deserializer::from_slice(&message);
+            let info = Message::deserialize(&mut de)?;
+
+            Ok(info)
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("read: ")]
+    Read(#[from] upgrade::ReadOneError),
```

Comment:
With the added error handling for the write part, this is a bit weird because write just emits and `io::Error` and the `io::Error` is part of this one.

Can you figure something out here that makes sense? Potentially flattening the error cases maybe?